src/chunker/disprove_execution.rs

```diff
@@ -0,0 +1,301 @@
+use std::rc::Rc;
+
+use ark_groth16::{Proof, VerifyingKey};
+
+use super::{
+    assigner::BCAssigner, chunk_groth16_verifier::groth16_verify_to_segments, common::RawWitness,
+    elements::dummy_element,
+};
+
+#[derive(Clone)]
+pub struct RawProof {
+    pub proof: Proof<ark_bn254::Bn254>,
+    pub public: Vec<<ark_bn254::Bn254 as ark_ec::pairing::Pairing>::ScalarField>,
+    pub vk: VerifyingKey<ark_bn254::Bn254>,
+}
+
+pub fn disprove_exec<A: BCAssigner>(
+    assigner: &A,
+    assert_witness: Vec<Vec<RawWitness>>,
+    wrong_proof: RawProof,
+) -> Option<(usize, RawWitness)> {
+    // 0. TODO: if 'wrong_proof' is correct, return none
+
+    // 1. derive assigner from wrong proof
+    let mut wrong_proof_assigner = A::default();
+    let mut segments = groth16_verify_to_segments(
+        &mut wrong_proof_assigner,
+        &wrong_proof.public,
+        &wrong_proof.proof,
+        &wrong_proof.vk,
+    );
+    let segment_length = segments.len();
+
+    // 2. recover assigner from witness
+    let hash_map = assigner.recover_from_witness(assert_witness);
+
+    // 3. find which chunk is unconsistent
+    for (idx, segment) in segments.iter_mut().enumerate() {
+        let mut is_param_equal = true;
+        for param in segment.parameter_list.iter() {
+            if param.to_hash().unwrap() != hash_map.get(param.id()).unwrap().clone() {
+                is_param_equal = false;
+            }
+        }
+        let mut is_result_equal = true;
+        for result in segment.result_list.iter_mut() {
+            if result.to_hash().unwrap() != hash_map.get(result.id()).unwrap().clone() {
+                is_result_equal = false;
+                // replace the result to hash_map
+                *result = Rc::new(Box::new(dummy_element(
+                    &result.id(),
+                    hash_map.get(result.id()).unwrap().clone(),
+                )));
+            }
+        }
+
+        if is_param_equal && !is_result_equal {
+            let disprove_witness = segment.witness(assigner);
+            return Some((idx, disprove_witness));
+        }
+    }
+
+    // if all intermediate values is identical, then return the final chunk
+    for (idx, segment) in segments.iter().enumerate() {
+        if segment.is_final() {
+            return Some((idx, segment.witness(assigner)));
+        }
+    }
+
+    panic!("shouldn't be here")
```

Comment:
Can you clean up the todo, comments and panic in this function?

Reply:
done